File: skills/agent/agent-doorbell/scripts/hook-runner.py
#!/usr/bin/env python3
import argparse
import json
import subprocess
import sys
from pathlib import Path


REASON_BY_EVENT = {
    "Elicitation": "needs-input",
    "PermissionRequest": "needs-input",
    "StopFailure": "blocked",
    "TeammateIdle": "done",
}


def limit_text(text, max_length):
    text = (text or "").strip()
    if len(text) <= max_length:
        return text
    return text[: max(0, max_length - 3)] + "..."


def read_event():
    raw = sys.stdin.read()
    if not raw.strip():
        return {}
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return {"hook_event_name": "unknown"}


def infer_agent_name(event):
    if event.get("teammate_name"):
        return event["teammate_name"]
    if event.get("agent_type"):
        return event["agent_type"]
    if event.get("tool_name"):
        return f"Agent permission: {event['tool_name']}"
    if event.get("mcp_server_name"):
        return f"Agent input: {event['mcp_server_name']}"
    return "Agent"


def infer_reason(event):
    event_name = event.get("hook_event_name", "")
    if event_name in REASON_BY_EVENT:
        return REASON_BY_EVENT[event_name]

    message = (event.get("last_assistant_message") or "").lower()
    if any(token in message for token in ("blocked", "can't proceed", "cannot proceed", "unable to proceed")):
        return "blocked"
    if any(token in message for token in ("cancelled", "canceled", "stopped by user")):
        return "cancelled"
    if any(token in message for token in ("?", "confirm", "approval", "permission", "choose", "which option")):
        return "needs-input"
    return "done"


def infer_summary(event, reason):
    event_name = event.get("hook_event_name") or "Agent stop"
    if event_name == "PermissionRequest":
        tool_name = event.get("tool_name") or "tool"
        return limit_text(f"Waiting for permission to use {tool_name}", 180)
    if event_name == "Elicitation":
        message = event.get("message") or "external input"
        return limit_text(f"Waiting for {message}", 180)
    if event_name == "StopFailure":
        error = event.get("error") or "error"
        return limit_text(f"Stopped because of {error}", 180)
    if event_name == "SubagentStop":
        message = event.get("last_assistant_message")
        if message:
            return limit_text(message, 180)
        return "Subagent stopped"
    if event_name == "TeammateIdle":
        return "Teammate became idle"
    if reason == "needs-input":
        return "Stopped and waiting for user input"
    if reason == "blocked":
        return "Stopped because progress is blocked"
    if reason == "cancelled":
        return "Stopped after cancellation or redirect"
    message = event.get("last_assistant_message")
    if message:
        return limit_text(message, 180)
    return "Stopped and handed control back"


def parse_args():
    parser = argparse.ArgumentParser(description="Run the Agent Doorbell cue from a lifecycle hook.")
    parser.add_argument("--mode", choices=("auto", "toast", "sound", "both", "none"), default="both")
    parser.add_argument("--intensity", choices=("quiet", "normal", "loud"), default="normal")
    parser.add_argument("--dry-run", action="store_true")
    return parser.parse_args()


def main():
    args = parse_args()
    event = read_event()
    reason = infer_reason(event)
    agent_name = limit_text(infer_agent_name(event), 48)
    summary = infer_summary(event, reason)
    ring_script = Path(__file__).resolve().with_name("ring.py")

    command = [
        sys.executable,
        str(ring_script),
        "--agent-name",
        agent_name,
        "--summary",
        summary,
        "--reason",
        reason,
        "--mode",
        args.mode,
        "--intensity",
        args.intensity,
    ]

    if args.dry_run:
        print(json.dumps({
            "event": event.get("hook_event_name"),
            "agent": agent_name,
            "reason": reason,
            "summary": summary,
            "command": command,
        }, ensure_ascii=False, separators=(",", ":")))
        return 0

    try:
        subprocess.run(command, check=False, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    except Exception:
        pass

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/agent/agent-doorbell/scripts/hook-runner.js
#!/usr/bin/env node
const { spawn } = require("child_process");
const fs = require("fs");
const path = require("path");

const REASON_BY_EVENT = {
  AfterAgent: "done",
  Elicitation: "needs-input",
  Notification: "needs-input",
  PermissionRequest: "needs-input",
  StopFailure: "blocked",
  TeammateIdle: "done",
};

function parseArgs(argv) {
  const args = { mode: "both", intensity: "normal", output: "none", dryRun: false };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    const value = argv[index + 1];
    if (arg === "--mode") {
      args.mode = value || args.mode;
      index += 1;
    } else if (arg === "--intensity") {
      args.intensity = value || args.intensity;
      index += 1;
    } else if (arg === "--output") {
      args.output = value || args.output;
      index += 1;
    } else if (arg === "--dry-run") {
      args.dryRun = true;
    }
  }
  return args;
}

function limitText(text, maxLength) {
  const value = String(text || "").trim();
  if (value.length <= maxLength) return value;
  return `${value.slice(0, Math.max(0, maxLength - 3))}...`;
}

function readEvent() {
  let raw = "";
  try {
    raw = fs.readFileSync(0, "utf8");
  } catch (_error) {
    return {};
  }
  if (!raw.trim()) return {};
  try {
    return JSON.parse(raw);
  } catch (_error) {
    return { hook_event_name: "unknown" };
  }
}

function eventMessage(event) {
  for (const key of ["last_assistant_message", "prompt_response", "message", "summary", "content"]) {
    const value = event[key];
    if (typeof value === "string" && value.trim()) return value.trim();
  }
  return "";
}

function inferAgentName(event) {
  if (event.agent_name) return event.agent_name;
  if (event.teammate_name) return event.teammate_name;
  if (event.agent_type) return event.agent_type;
  if (event.tool_name) return `Agent permission: ${event.tool_name}`;
  if (event.mcp_server_name) return `Agent input: ${event.mcp_server_name}`;
  return "Agent";
}

function inferReason(event) {
  const eventName = event.hook_event_name || "";
  if (REASON_BY_EVENT[eventName]) return REASON_BY_EVENT[eventName];

  const message = eventMessage(event).toLowerCase();
  if (["blocked", "can't proceed", "cannot proceed", "unable to proceed"].some((token) => message.includes(token))) {
    return "blocked";
  }
  if (["cancelled", "canceled", "stopped by user"].some((token) => message.includes(token))) {
    return "cancelled";
  }
  if (["?", "confirm", "approval", "permission", "choose", "which option"].some((token) => message.includes(token))) {
    return "needs-input";
  }
  return "done";
}

function inferSummary(event, reason) {
  const eventName = event.hook_event_name || "Agent stop";
  if (eventName === "AfterAgent") return limitText(eventMessage(event) || "Agent finished responding", 180);
  if (eventName === "Notification") return limitText(eventMessage(event) || "Agent notification needs attention", 180);
  if (eventName === "PermissionRequest") return limitText(`Waiting for permission to use ${event.tool_name || "tool"}`, 180);
  if (eventName === "Elicitation") return limitText(`Waiting for ${event.message || "external input"}`, 180);
  if (eventName === "StopFailure") return limitText(`Stopped because of ${event.error || "error"}`, 180);
  if (eventName === "SubagentStop") return limitText(eventMessage(event) || "Subagent stopped", 180);
  if (eventName === "TeammateIdle") return "Teammate became idle";
  if (reason === "needs-input") return "Stopped and waiting for user input";
  if (reason === "blocked") return "Stopped because progress is blocked";
  if (reason === "cancelled") return "Stopped after cancellation or redirect";
  return limitText(eventMessage(event) || "Stopped and handed control back", 180);
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const event = readEvent();
  const reason = inferReason(event);
  const agentName = limitText(inferAgentName(event), 48);
  const summary = inferSummary(event, reason);
  const ringScript = path.join(__dirname, "ring.js");
  const command = [
    process.execPath,
    ringScript,
    "--agent-name",
    agentName,
    "--summary",
    summary,
    "--reason",
    reason,
    "--mode",
    args.mode,
    "--intensity",
    args.intensity,
  ];

  if (args.dryRun) {
    console.log(JSON.stringify({
      event: event.hook_event_name,
      agent: agentName,
      reason,
      summary,
      command,
    }));
    return;
  }

  try {
    const child = spawn(command[0], command.slice(1), {
      detached: true,
      stdio: "ignore",
      windowsHide: true,
    });
    child.unref();
  } catch (_error) {
    // Hooks must never block or fail the Agent response path.
  }

  if (args.output === "gemini") {
    console.log(JSON.stringify({ suppressOutput: true }));
  }
}

main();
